src/Ic/DiscIc.cpp

```diff
@@ -0,0 +1,250 @@
+
+#include <cstdlib>
+#include <cmath>
+#include "Precision.h"
+#include "Debug.h"
+#include "Ic.h"
+
+
+//=================================================================================================
+//  DiscIc::DiscIc
+/// Class constructor
+//=================================================================================================
+template <int ndim>
+DiscIc<ndim>::DiscIc(Simulation<ndim>* _sim, FLOAT _invndim) :
+  Ic<ndim>(_sim, _invndim)
+{
+  if (simparams->intparams["dimensionless"] == 0) {
+    ExceptionHandler::getIstance().raise("dimensionless units required");
+  }
+  if (simparams->intparams["ndim"] == 1){
+    ExceptionHandler::getIstance().raise("The disc IC require at least 2 dimensions!");
+  }
+  if (simparams->intparams["sink_particles"] == 0) {
+    ExceptionHandler::getIstance().raise("You can't run a disc simulation without sink accretion!");
+  }
+}
+
+//=================================================================================================
+//  DiscIc::Generate
+/// Generate an accretion disc
+//=================================================================================================
+template <int ndim>
+void DiscIc<ndim>::Generate(void)
+{
+  debug2("[DiscIc::Generate");
+
+  if (simparams->stringparams["gas_eos"] != "disc_locally_isothermal") {
+    cout << "Warning: you have not selected the disc locally isothermal EOS!" << endl;
+    cout << "Are you sure this is what you want?" << endl;
+  }
+
+  // Local copy of important parameters
+  const int Npart = simparams->intparams["Nhydro"];
+  const FLOAT gammaone = simparams->floatparams["gamma_eos"] - 1.0;
+  const FLOAT mass = simparams->floatparams["DiscIcMass"];
+  const FLOAT p = simparams->floatparams["DiscIcP"];
+  const FLOAT q = simparams->floatparams["DiscIcQ"];
+  const FLOAT rin = simparams->floatparams["DiscIcRin"];
+  const FLOAT rout = simparams->floatparams["DiscIcRout"];
+  const FLOAT H_r = simparams->floatparams["DiscIcHr"];
+  const bool HaveDust = (simparams->stringparams["dust_forces"] != "none");
+  int NDust = 0;
+  int Ntotal = Npart;
+  if (HaveDust) {
+    NDust = simparams->intparams["Ndust"];
+    Ntotal += NDust;
+  }
+  const FLOAT DustGasRatio = simparams->floatparams["DustGasRatio"];
+
+  // Assume units where GM_\ast=1
+  const FLOAT GStar_M = 1;
+
+
+  // Allocate global and local memory for all particles
+  hydro->Nhydro = Ntotal;
+  if (simparams->intparams["DiscIcPlanet"]==0) {
+    sim->nbody->Nstar = 1;
+  }
+  else {
+    sim->nbody->Nstar = 2;
+  }
+
+  sim->AllocateParticleMemory();
+
+  // Compute the particle mass
+  FLOAT massi = mass/Npart;
+  if (HaveDust) massi *= 1-DustGasRatio;
+  const FLOAT massi_dust = mass*DustGasRatio/NDust;
+
+
+  // Normalisation of sound speed at rin
+  const FLOAT cs0 = H_r*std::sqrt(GStar_M/rin);
+
+  // Normalization of the gas surface density at rin
+  const FLOAT sig0=(2-p)*mass*(1-DustGasRatio)*std::pow(rin,-p)/(2*pi) / (pow(rout,2-p)-pow(rin,2-p));
```

Comment:
This will fail for  p=2. Can we either have a sanity check or the necessary code to support it?